server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java

```diff
@@ -53,17 +54,75 @@ public static void validateMapperStructure(Collection<ObjectMapper> objectMapper
-                } else if (fieldNames.add(name) == false) {
-                    throw new IllegalArgumentException("Field [" + name + "] is defined twice.");
-                }
-            });
+        for (FieldMapper fieldMapper : fieldMappers) {
+            String name = fieldMapper.name();
+            if (objectFullNames.contains(name)) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined both as an object and a field.");
+            } else if (fieldNames.add(name) == false) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined twice.");
+            }
+
+            validateFieldMapper(fieldMapper, fieldTypes);
+        }
+
+        Set<String> fieldAliasNames = new HashSet<>();
+        for (FieldAliasMapper fieldAliasMapper : fieldAliasMappers) {
+            String name = fieldAliasMapper.name();
+            if (objectFullNames.contains(name)) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined both as an object and a field.");
+            } else if (fieldNames.contains(name)) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined both as an alias and a concrete field.");
+            } else if (fieldAliasNames.add(name) == false) {
+                throw new IllegalArgumentException("Field [" + name + "] is defined twice.");
+            }
+
+            validateFieldAliasMapper(name, fieldAliasMapper.path(), fieldNames, fieldAliasNames);
+        }
     }
 
+    /**
+     * Checks that the new field mapper does not conflict with existing mappings.
+     */
+    private static void validateFieldMapper(FieldMapper fieldMapper,
+                                            FieldTypeLookup fieldTypes) {
+        MappedFieldType newFieldType = fieldMapper.fieldType();
+        MappedFieldType existingFieldType = fieldTypes.get(newFieldType.name());
+
+        if (existingFieldType != null && Objects.equals(newFieldType, existingFieldType) == false) {
+            List<String> conflicts = new ArrayList<>();
+            existingFieldType.checkCompatibility(newFieldType, conflicts);
+            if (conflicts.isEmpty() == false) {
+                throw new IllegalArgumentException("Mapper for [" + newFieldType.name() +
+                    "] conflicts with existing mapping:\n" + conflicts.toString());
+            }
+        }
+    }
+
+    /**
+     * Checks that the new field alias is valid.
+     *
+     * Note that this method assumes that new concrete fields have already been processed, so that it
+     * can verify that an alias refers to an existing concrete field.
+     */
+    private static void validateFieldAliasMapper(String aliasName,
+                                                 String path,
+                                                 Set<String> fieldMappers,
+                                                 Set<String> fieldAliasMappers) {
```

Comment:
similar here, is there a reason why we omitted the check that a new alias name doesn't conflict with existing field names?

Reply:
Same comment as above -- it is now covered by the if/ else checks in `validateNewMappers`.